golem-worker-executor-base/src/services/plugins.rs

```diff
@@ -0,0 +1,476 @@
+// Copyright 2024 Golem Cloud
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::error::GolemError;
+use crate::grpc::{authorised_grpc_request, UriBackConversion};
+use crate::services::golem_config::ComponentServiceConfig;
+use async_trait::async_trait;
+use golem_api_grpc::proto::golem::component::v1::component_service_client::ComponentServiceClient;
+use golem_api_grpc::proto::golem::component::v1::plugin_service_client::PluginServiceClient;
+use golem_api_grpc::proto::golem::component::v1::{
+    get_installed_plugins_response, get_plugin_response, GetInstalledPluginsRequest,
+    GetPluginRequest,
+};
+use golem_common::cache::{BackgroundEvictionMode, Cache, FullCacheEvictionMode, SimpleCache};
+use golem_common::client::{GrpcClient, GrpcClientConfig};
+use golem_common::config::RetryConfig;
+use golem_common::model::plugin::{
+    DefaultPluginOwner, DefaultPluginScope, PluginDefinition, PluginInstallation, PluginOwner,
+    PluginScope,
+};
+use golem_common::model::{AccountId, ComponentId, ComponentVersion, PluginInstallationId};
+use http::Uri;
+use std::sync::Arc;
+use tonic::codec::CompressionEncoding;
+use tonic::transport::Channel;
+use uuid::Uuid;
+
+/// Part of the `Plugins` service for recording observed information as a way to pre-cache
+/// data. It is in a separate trait because it does not have to be parametric for the Owner/Scope
+/// types.
+#[async_trait]
+pub trait PluginsObservations {
```

Comment:
Do we need to care about invalidating this cache or will we always end up calling observe_plugin_installation with uncached data when (re) running a worker?

Reply:
The installations are immutable so no need for invalidation. (And old entries will be evicted when the cache is full)